### PR TITLE
feat: support disable property for json schema according to APISIX's change

### PR DESCRIPTION
**Pull request by nic-chen**
The Apache APISIX Dashboard is using a module named  [react-jsonschema-form](https://rjsf-team.github.io/react-jsonschema-form/), which is used to build UI components according to JSONSchema. 

But due to some differences between Draft7 and other versions Apache APISIX is using, some components couldn't be rendered correctly.

We need to support `disable` property for JSONSchema for those who has no `properties` or empty `properties`, it's what this PR resolved.

- Related PR
https://github.com/apache/apisix/pull/2099

- Related issues
https://github.com/apache/apisix/issues/2082
#495

### CODE REVIEW

**Comment by LiteSun** [comment]
Need to deal with conflict. @nic-chen

**Comment by nic-chen** [comment]
> Need to deal with conflict. @nic-chen

Thanks for remind.

**Comment by membphis** [comment]
@gxthrj @ShiningRush @tokers pls take a look at this PR when you have time

File: api/internal/core/store/validate.go [hunk 1]
```diff
@@ -238,22 +238,37 @@ func (v *APISIXJsonSchemaValidator) Validate(obj interface{}) error {
...
+					schemaMap["properties"] = pluginDisableSchema
+				}
+			} else {
+				schemaMap["properties"] = pluginDisableSchema
+			}
+		}
+		schemaByte, err := json.Marshal(schemaMap)
+		if err != nil {
+			log.Warnf("scheme validate failed: schema not found, path: %s, %w", "plugins."+pluginName, err)
+			return fmt.Errorf("scheme validate failed: schema not found, path: %s, %w", "plugins."+pluginName, err)
```
Comment (ShiningRush) [inline api/internal/core/store/validate.go:268]: error info is wrong, here is a json.Marshal action.

File: api/internal/core/store/validate.go [hunk 2]
```diff
@@ -238,22 +238,37 @@ func (v *APISIXJsonSchemaValidator) Validate(obj interface{}) error {
...
 		}
-		if schemaDef == "" {
-			log.Warnf("schema validate failed: schema not found, path: %s", "plugins."+pluginName)
+		if schemaValue == nil {
+			log.Warnf("schema validate failed: schema not found,  %s, %s", "plugins."+pluginName, schemaType)
 			return fmt.Errorf("schema validate failed: schema not found, path: %s", "plugins."+pluginName)
 		}
+		schemaMap := schemaValue.(map[string]interface{})
+		if schemaMap["type"] != nil && schemaMap["type"].(string) == "object" {
+			if properties, ok := schemaMap["properties"]; ok {
```
Comment (ShiningRush) [inline api/internal/core/store/validate.go:256]: Too much branch this logic, we can refactor to
```go
// guard code
if schemaMap["properties"] == nil {
    schemaMap["properties"]  = make(map[string]interface{})
}
// ensure
if len(schemaMap["properties"].(map[string]interface{})) == 0 {
  schemaMap["properties"] = pluginDisableSchema
}
```
BTW: I am curious if it is expected logic that we do nothing when user specified `properties` even it does not contain `disable`?

File: api/internal/core/store/validate.go [hunk 3]
```diff
@@ -238,22 +238,37 @@ func (v *APISIXJsonSchemaValidator) Validate(obj interface{}) error {
...
+				if len(propertiesMap) == 0 {
+					schemaMap["properties"] = pluginDisableSchema
+				}
+			} else {
+				schemaMap["properties"] = pluginDisableSchema
+			}
+		}
+		schemaByte, err := json.Marshal(schemaMap)
+		if err != nil {
+			log.Warnf("scheme validate failed: schema not found, path: %s, %w", "plugins."+pluginName, err)
```
Comment (idbeta) [inline api/internal/core/store/validate.go:267]: `scheme` -> `schema`

File: api/internal/core/store/validate.go (hunk 1 above)
Comment (nic-chen) [reply api/internal/core/store/validate.go]: updated.

File: api/internal/core/store/validate.go (hunk 2 above)
Comment (nic-chen) [reply api/internal/core/store/validate.go]: @ShiningRush  we only care about `properties` is nil or empty. please refer: https://github.com/apache/apisix/pull/2099/files

File: api/internal/core/store/validate.go (hunk 3 above)
Comment (nic-chen) [reply api/internal/core/store/validate.go]: fixed.

**Comment by juzhiyuan** [comment]
update?

**Comment by nic-chen** [comment]
@ShiningRush please review again，thanks.

**Comment by nic-chen** [comment]
@gxthrj @tokers please help review at your convenience，thanks.

File: api/internal/core/store/validate_test.go [hunk 4]
```diff
@@ -249,6 +248,30 @@ func TestAPISIXJsonSchemaValidator_checkUpstream(t *testing.T) {
 	assert.EqualError(t, err, "schema validate failed: (root): Does not match pattern '^((uri|server_name|server_addr|request_uri|remote_port|remote_addr|query_string|host|hostname)|arg_[0-9a-zA-z_-]+)$'")
 }
 
+func TestAPISIXJsonSchemaValidator_Plugin(t *testing.T) {
+	validator, err := NewAPISIXJsonSchemaValidator("main.route")
+	assert.Nil(t, err)
+
+	// plugin empty schema
+	route := &entity.Route{}
+	reqBody := `{
```
Comment (moonming) [inline api/internal/core/store/validate_test.go:257]: Can't understand the test case

**Comment by moonming** [comment]
What is this pr used for?  I can not understand it too

**Comment by nic-chen** [comment]
> What is this pr used for? I can not understand it too

updated the description. please have a look.

File: api/internal/core/store/validate_test.go [hunk 5]
```diff
@@ -249,6 +248,30 @@ func TestAPISIXJsonSchemaValidator_checkUpstream(t *testing.T) {
 	assert.EqualError(t, err, "schema validate failed: (root): Does not match pattern '^((uri|server_name|server_addr|request_uri|remote_port|remote_addr|query_string|host|hostname)|arg_[0-9a-zA-z_-]+)$'")
 }
 
+func TestAPISIXJsonSchemaValidator_Plugin(t *testing.T) {
+	validator, err := NewAPISIXJsonSchemaValidator("main.route")
+	assert.Nil(t, err)
+
+	// validate plugin's schema which has no `properties` or empty `properties`
```
Comment (membphis) [inline api/internal/core/store/validate_test.go:255]: need E2E test cases @nic-chen

**Review by membphis — CHANGES_REQUESTED** [review]

File: api/internal/core/store/validate_test.go (hunk 5 above)
Comment (nic-chen) [reply api/internal/core/store/validate_test.go:255]: @membphis  added.

Comment (ShiningRush) [reply api/internal/core/store/validate_test.go:255]: Hmmm, I think we can not put all test case to e2e test, e2e test is too heavy to maintain.So e2e usually used for ensuring key path of the software, If some cases could covered by unit test, just maintain unit test cases is fine.

File: api/internal/core/store/validate_test.go [hunk 6]
```diff
@@ -249,6 +248,30 @@ func TestAPISIXJsonSchemaValidator_checkUpstream(t *testing.T) {
 	assert.EqualError(t, err, "schema validate failed: (root): Does not match pattern '^((uri|server_name|server_addr|request_uri|remote_port|remote_addr|query_string|host|hostname)|arg_[0-9a-zA-z_-]+)$'")
 }
 
+func TestAPISIXJsonSchemaValidator_Plugin(t *testing.T) {
```
Comment (ShiningRush) [inline api/internal/core/store/validate_test.go:251]: This case coverage is too low, it does not cover any failed case : (

File: api/internal/core/store/validate_test.go (hunk 5 above)
Comment (nic-chen) [reply api/internal/core/store/validate_test.go:255]: @ShiningRush  we need to ensure the DP works well with `disable` property too.

**Review by membphis — APPROVED** [review]
read the E2E test cases, this PR is good for me.

need @ShiningRush @gxthrj to confirm the Golang source code

**Review by ShiningRush — APPROVED** [review]

**Pull request merged**